s3stream/src/main/java/com/automq/stream/s3/operator/DeleteObjectsAccumulator.java

```diff
@@ -70,93 +58,79 @@ public PendingDeleteRequest(List<ObjectStorage.ObjectPath> deleteObjectPath, Com
-
-            return null;
-        });
-    }
-
-    public void run() {
-        ArrayList<ObjectStorage.ObjectPath> deleteKeys = new ArrayList<>(1000);
-        ArrayList<CompletableFuture<Void>> futures = new ArrayList<>();
-
-        while (running.get()) {
-            PendingDeleteRequest item;
-            try {
-                item = deleteBatchQueue.poll(500, TimeUnit.MICROSECONDS);
-                if (item == null) {
-                    if (!deleteKeys.isEmpty()) {
-                        submitDeleteObjectsRequest(deleteKeys, futures);
-                        deleteKeys = new ArrayList<>(1000);
-                        futures = new ArrayList<>();
-                    }
-
-                    continue;
+        if (!concurrentRequestLimiter.tryAcquire()) {
+            deleteRequestQueue.add(new PendingDeleteRequest(objectPaths, subBatchCf));
+            return;
+        }
+        deleteObjectsFunction.apply(objectKeys)
+            .whenComplete((res, e) -> concurrentRequestLimiter.release())
+            .thenAccept(nil -> {
+                LOGGER.info("Delete objects finished, count: {}, cost: {}ms", objectKeys.size(), timerUtil.elapsedAs(TimeUnit.MILLISECONDS));
+                S3OperationStats.getInstance().deleteObjectsStats(true).record(timerUtil.elapsedAs(TimeUnit.NANOSECONDS));
+                subBatchCf.complete(null);
+                handleDeleteRequestQueue();
+            }).exceptionally(ex -> {
+                if (ex instanceof AbstractObjectStorage.DeleteObjectsException) {
+                    AbstractObjectStorage.DeleteObjectsException deleteObjectsException = (AbstractObjectStorage.DeleteObjectsException) ex;
+                    LOGGER.warn("Delete objects failed, count: {}, cost: {}, failedKeys: {}",
+                        deleteObjectsException.getFailedKeys().size(), timerUtil.elapsedAs(TimeUnit.NANOSECONDS),
+                        deleteObjectsException.getFailedKeys());
+                } else {
+                    S3OperationStats.getInstance().deleteObjectsStats(false)
+                        .record(timerUtil.elapsedAs(TimeUnit.NANOSECONDS));
+                    LOGGER.info("Delete objects failed, count: {}, cost: {}, ex: {}",
+                        objectKeys.size(), timerUtil.elapsedAs(TimeUnit.NANOSECONDS), ex.getMessage());
                 }
+                subBatchCf.completeExceptionally(ex);
+                handleDeleteRequestQueue();
+                return null;
+            });
+    }
 
-                // enqueue batch
-                deleteKeys.addAll(item.deleteObjectPath);
-                futures.add(item.future);
-
-                // submit batch
-                if (deleteKeys.size() >= this.maxBatchSize) {
-                    submitDeleteObjectsRequest(deleteKeys, futures);
-                    deleteKeys = new ArrayList<>(1000);
-                    futures = new ArrayList<>();
+    private void handleDeleteRequestQueue() {
```

Comment:
Hi, I think this handle pendingRequest should be check batched.
if the queuing request number is 100, 100, 300 they can be submit one delete request to delete them all.